blog/2018-08-30-streaming-mysql-data-changes-into-kinesis.adoc

```diff
@@ -0,0 +1,411 @@
+
+The next step is to build an `EmbeddedEngine` instance from the configuration.
+Again this is done using a fluent API:
+
+[source,java]
+----
+EmbeddedEngine engine = EmbeddedEngine.create()
+    .using(config)
+    .using(this.getClass().getClassLoader())
+    .using(Clock.SYSTEM)
+    .notifying(this::sendRecord)
+    .build();
+----
+
+The most interesting part here is the `notifying` call.
+The method passed here is the one which will be invoked by the engine for each emitted data change record.
+So let's take a look at the implementation of this method.
+
+== Sending Change Records to Kinesis
+
+The `sendRecord()` method is where the magic happens.
+We'll convert the incoming `SourceRecord` into an equivalent JSON representation and propagate it to a Kinesis stream.
+
+For that, it's important to understand some conceptual differences between Apache Kafka and Kinesis.
+Specifically, messages in Kafka have a _key_ and a _value_ (which both are arbitrary byte arrays).
+In case of Debezium, the key of data change events represents the primary key of the affected record and the value is a structure comprising of old and new row state as well as some additional metadata.
+
+In Kinesis on the other hand a message contains a _data blob_ (again an arbitrary byte sequence) and a _partition key_.
+Kinesis streams can be split up into multiple shards and the partition key is used to determine into which shard a given message should go.
+
+Now one could think of mapping the key from Debezium's change data events to the Kinesis partition key,
+but partition keys are limited to a length of 256 bytes.
+Depending on the length of primary key column(s) in the captured tables, this might not be enough.
+So a safer option is to create a hash value from the change message key and use that as the partition key.
+This in turn means that the change message key structure should be added next to the actual value to the Kinesis message's data blob.
+While the key column values themselves are part of the value structure, too, a consumer otherwise wouldn't know which column(s) make up the primary key.
+
+With that in mind, let's take a look at the `sendRecord()` implementation:
+
+[source,java]
+----
+private void sendRecord(SourceRecord record) {
+    // We are interested only in data events not schema change events
+    if (record.topic().equals("kinesis")) {
+        return;
+    }
+
+    // create schema for container with key *and* value
+    Schema schema = SchemaBuilder.struct()
+        .field("key", record.keySchema())
+        .field("value", record.valueSchema())
+        .build();
+
+    Struct message = new Struct(schema);
+    message.put("key", record.key());
+    message.put("value", record.value());
+
+    // create partition key by hashing the record's key
+    String partitionKey = String.valueOf(record.key() != null ? record.key().hashCode() : new Object().hashCode());
```

Comment:
Should not all records without key end up in the same partition?

Reply:
Yes, good thinking. I'll set it to a fixed value.